File: judo-sdk/src/main/java/com/judopay/model/CheckCardRequest.java
package com.judopay.model;

import com.judopay.api.Request;
import com.judopay.error.JudoIdInvalidError;

import java.util.Map;

import static com.judopay.arch.TextUtil.isEmpty;
import static com.judopay.model.LuhnCheck.isValid;

/**
 * Represents the data needed to perform a check card transaction with the judo API.
 * Use the {@link CheckCardRequest.Builder} for object construction.
 * <p>
 * When creating a {@link CheckCardRequest} the {@link CheckCardRequest#judoId}
 * must be provided.
 */
@SuppressWarnings("unused")
public final class CheckCardRequest extends Request {
    private String judoId;
    private String yourConsumerReference;
    private Address cardAddress;
    private String cardNumber;
    private String cv2;
    private String expiryDate;
    private String startDate;
    private String emailAddress;
    private String mobileNumber;
    private String issueNumber;
    private String currency;
    private Map<String, String> yourPaymentMetaData;
    private PrimaryAccountDetails primaryAccountDetails;

    public String getJudoId() {
        return judoId;
    }

    public String getYourConsumerReference() {
        return yourConsumerReference;
    }

    public Address getCardAddress() {
        return cardAddress;
    }

    public String getCardNumber() {
        return cardNumber;
    }

    public String getCv2() {
        return cv2;
    }

    public String getExpiryDate() {
        return expiryDate;
    }

    public String getStartDate() {
        return startDate;
    }

    public String getIssueNumber() {
        return issueNumber;
    }

    public Map<String, String> getMetaData() {
        return yourPaymentMetaData;
    }

    public String getEmailAddress() {
        return emailAddress;
    }

    public String getMobileNumber() {
        return mobileNumber;
    }

    public String getCurrency() {
        return currency;
    }

    @SuppressWarnings("UnusedReturnValue")
    public static class Builder {
        private String judoId;
        private String consumerReference;
        private String paymentReference;
        private Address cardAddress;
        private String cardNumber;
        private String currency;
        private String cv2;
        private String expiryDate;
        private String startDate;
        private String issueNumber;
        private String emailAddress;
        private String mobileNumber;
        private Map<String, String> yourPaymentMetaData;
        private PrimaryAccountDetails primaryAccountDetails;

        public Builder setJudoId(final String judoId) {
            this.judoId = judoId;
            return this;
        }

        public Builder setConsumerReference(final String consumerReference) {
            this.consumerReference = consumerReference;
            return this;
        }

        public Builder setPaymentReference(final String paymentReference) {
            this.paymentReference = paymentReference;
            return this;
        }

        public Builder setCardAddress(final Address cardAddress) {
            this.cardAddress = cardAddress;
            return this;
        }

        public Builder setCardNumber(final String cardNumber) {
            this.cardNumber = cardNumber;
            return this;
        }

        public Builder setCv2(final String cv2) {
            this.cv2 = cv2;
            return this;
        }

        public Builder setCurrency(@Currency.Type final String currency) {
            this.currency = currency;
            return this;
        }

        public Builder setExpiryDate(final String expiryDate) {
            this.expiryDate = expiryDate;
            return this;
        }

        public Builder setStartDate(final String startDate) {
            this.startDate = startDate;
            return this;
        }

        public Builder setIssueNumber(final String issueNumber) {
            this.issueNumber = issueNumber;
            return this;
        }

        public Builder setEmailAddress(final String emailAddress) {
            this.emailAddress = emailAddress;
            return this;
        }

        public Builder setMobileNumber(final String mobileNumber) {
            this.mobileNumber = mobileNumber;
            return this;
        }

        public Builder setMetaData(final Map<String, String> metaData) {
            this.yourPaymentMetaData = metaData;
            return this;
        }

        public Builder setPrimaryAccountDetails(final PrimaryAccountDetails primaryAccountDetails) {
            this.primaryAccountDetails = primaryAccountDetails;
            return this;
        }

        public CheckCardRequest build() {
            if (isEmpty(judoId) || !isValid(judoId)) {
                throw new JudoIdInvalidError();
            }

            checkNotNull(cardNumber);
            checkNotNull(cv2);
            checkNotNull(expiryDate);

            CheckCardRequest request = new CheckCardRequest();

            request.judoId = judoId;
            request.currency = currency;
            request.yourConsumerReference = consumerReference;
            request.cardAddress = cardAddress;
            request.cardNumber = cardNumber;
            request.cv2 = cv2;
            request.expiryDate = expiryDate;
            request.startDate = startDate;
            request.issueNumber = issueNumber;
            request.emailAddress = emailAddress;
            request.mobileNumber = mobileNumber;
            request.yourPaymentMetaData = yourPaymentMetaData;
            request.primaryAccountDetails = primaryAccountDetails;

            if (!isEmpty(paymentReference)) {
                request.yourPaymentReference = paymentReference;
            }
            return request;
        }
    }
}

File: judokit-android/src/main/java/com/judopay/judokit/android/api/model/request/CheckCardRequest.kt
package com.judopay.judokit.android.api.model.request

import com.judopay.judokit.android.api.model.request.threedsecure.ThreeDSecureTwo
import com.judopay.judokit.android.model.ChallengeRequestIndicator
import com.judopay.judokit.android.model.PrimaryAccountDetails
import com.judopay.judokit.android.model.ScaExemption
import com.judopay.judokit.android.requireNotNull
import com.judopay.judokit.android.requireNotNullOrEmpty

class CheckCardRequest private constructor(
    private var uniqueRequest: Boolean?,
    private var yourPaymentReference: String,
    private var judoId: String,
    private var yourConsumerReference: String,
    private var cardAddress: Address?,
    private var cardNumber: String,
    private var cv2: String,
    private var expiryDate: String,
    private var startDate: String?,
    private var issueNumber: String?,
    private var currency: String?,
    private var primaryAccountDetails: PrimaryAccountDetails?,
    private var yourPaymentMetaData: Map<String, String>?,
    private var initialRecurringPayment: Boolean?,
    private var emailAddress: String?,
    private var mobileNumber: String?,
    private var phoneCountryCode: String?,
    private var threeDSecure: ThreeDSecureTwo?,
    private var cardHolderName: String?,
    private var challengeRequestIndicator: ChallengeRequestIndicator?,
    private var scaExemption: ScaExemption?,
    private var amount: String = "0",
) {
    class Builder {
        private var uniqueRequest: Boolean? = null
        private var yourPaymentReference: String? = null
        private var judoId: String? = null
        private var yourConsumerReference: String? = null
        private var address: Address? = null
        private var cardNumber: String? = null
        private var cv2: String? = null
        private var expiryDate: String? = null
        private var startDate: String? = null
        private var issueNumber: String? = null
        private var currency: String? = null
        private var primaryAccountDetails: PrimaryAccountDetails? = null
        private var yourPaymentMetaData: Map<String, String>? = null
        private var initialRecurringPayment: Boolean? = null
        private var threeDSecure: ThreeDSecureTwo? = null
        private var cardHolderName: String? = null
        private var phoneCountryCode: String? = null
        private var challengeRequestIndicator: ChallengeRequestIndicator? = null
        private var scaExemption: ScaExemption? = null
        private var emailAddress: String? = null
        private var mobileNumber: String? = null

        fun setUniqueRequest(uniqueRequest: Boolean?) = apply { this.uniqueRequest = uniqueRequest }

        fun setYourPaymentReference(yourPaymentReference: String?) =
            apply { this.yourPaymentReference = yourPaymentReference }

        fun setCurrency(currency: String?) = apply { this.currency = currency }

        fun setJudoId(judoId: String?) = apply { this.judoId = judoId }

        fun setYourConsumerReference(yourConsumerReference: String?) =
            apply { this.yourConsumerReference = yourConsumerReference }

        fun setYourPaymentMetaData(yourPaymentMetaData: Map<String, String>?) =
            apply { this.yourPaymentMetaData = yourPaymentMetaData }

        fun setAddress(address: Address?) = apply { this.address = address }

        fun setCardNumber(cardNumber: String?) = apply { this.cardNumber = cardNumber }

        fun setCv2(cv2: String?) = apply { this.cv2 = cv2 }

        fun setExpiryDate(expiryDate: String?) = apply { this.expiryDate = expiryDate }

        fun setStartDate(startDate: String?) = apply { this.startDate = startDate }

        fun setIssueNumber(issueNumber: String?) = apply { this.issueNumber = issueNumber }

        fun setPrimaryAccountDetails(primaryAccountDetails: PrimaryAccountDetails?) =
            apply { this.primaryAccountDetails = primaryAccountDetails }

        fun setInitialRecurringPayment(initialRecurringPayment: Boolean?) =
            apply { this.initialRecurringPayment = initialRecurringPayment }

        fun setThreeDSecure(threeDSecureTwo: ThreeDSecureTwo?) = apply { this.threeDSecure = threeDSecureTwo }

        fun setCardHolderName(cardHolderName: String?) =
            apply { this.cardHolderName = cardHolderName }

        fun setPhoneCountryCode(phoneCountryCode: String?) =
            apply { this.phoneCountryCode = phoneCountryCode }

        fun setChallengeRequestIndicator(challengeRequestIndicator: ChallengeRequestIndicator?) =
            apply { this.challengeRequestIndicator = challengeRequestIndicator }

        fun setScaExemption(scaExemption: ScaExemption?) =
            apply { this.scaExemption = scaExemption }

        fun setEmailAddress(emailAddress: String?) = apply { this.emailAddress = emailAddress }

        fun setMobileNumber(mobileNumber: String?) = apply { this.mobileNumber = mobileNumber }

        fun build(): CheckCardRequest {
            val id = requireNotNullOrEmpty(judoId, "judoId")
            val myCurrency = requireNotNullOrEmpty(currency, "currency")
            val consumerReference =
                requireNotNullOrEmpty(yourConsumerReference, "yourConsumerReference")
            val myCardNumber = requireNotNullOrEmpty(cardNumber, "cardNumber")
            val myCv2 = requireNotNullOrEmpty(cv2, "cv2")
            val myExpiryDate = requireNotNullOrEmpty(expiryDate, "expiryDate")
            val paymentReference =
                requireNotNullOrEmpty(yourPaymentReference, "yourPaymentReference")
            val myThreeDSecure = requireNotNull(threeDSecure, "threeDSecure")

            return CheckCardRequest(
                uniqueRequest,
                paymentReference,
                id,
                consumerReference,
                address,
                myCardNumber,
                myCv2,
                myExpiryDate,
                startDate,
                issueNumber,
                myCurrency,
                primaryAccountDetails,
                yourPaymentMetaData,
                initialRecurringPayment,
                emailAddress,
                mobileNumber,
                phoneCountryCode,
                myThreeDSecure,
                cardHolderName,
                challengeRequestIndicator,
                scaExemption
            )
        }
    }
}
